### ISSUE TITLE
Multi-Arch image

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/docker-images#102]
Am interested in seeing how we can standardize the images here a bit and consolidate maintenance effort. In particular am curious how we can make a multi-arch image. There's [a good blogpost]( https://blog.docker.com/2017/11/multi-arch-all-the-things/ ) on this topic. It would be good to determine what the current issues prompting the different images we have are and how we address those issues in a general way that benefits all of the images here.

**Comment by westurner** [comment]
Could this include the arch string in the miniforge download URL?

https://github.com/conda-forge/docker-images/blob/8ae623a13166aebad9a3bcbad82253436aefc58f/miniforge3/Dockerfile#L25


https://github.com/conda-forge/miniforge/releases
- [x] `Miniforge3-${MINIFORGE_VERSION}-Linux-x86_64.sh`
- [ ] `Miniforge3-${MINIFORGE_VERSION}-Linux-aarch64.sh`
- [ ] `Miniforge3-${MINIFORGE_VERSION}-Linux-ppe64le.sh`

How to make the Dockerfile pass the arch through to be templated into the miniforge release URL?

- [In docker experimental mode, the `--platform` flag works](https://github.com/moby/moby/issues/36552)
  - it sets [`DEB_TARGET_GNU_CPU=aarch64`](https://github.com/moby/moby/issues/36552#issuecomment-582051469)
- In lieu of waiting for things to emerge from experimental mode in the moby project, specifying the CPU arch as a build `ARG` or `ENV` variable (`TARGET_CPU_ARCH=x86_64 | aarch64 | ppe64le`  may be the most portable way to support multiple architectures with the miniforge docker container). 

...

- If multi-OS is also a goal, the ansible molecule dockerfile shows one way to execute different package commands depending upon the existence of various package managers:
  https://github.com/ansible-community/molecule/blob/master/molecule/data/Dockerfile.j2#L17-L22

**Comment by jakirkham** [comment]
That seems like a question to raise on [the miniforge issue tracker]( https://github.com/conda-forge/miniforge/issues ).

**Comment by westurner** [comment]
FWIU, there's nothing that needs to change in miniforge to make this happen.
The question is how to specify the TARGET_CPU_ARCH for builds of the miniforge Dockerfile that's in this project (conda-forge/docker-images).

https://github.com/conda-forge/docker-images/blob/8ae623a13166aebad9a3bcbad82253436aefc58f/miniforge3/Dockerfile#L25

Multi-arch images (including miniforge images) could be created with `docker build --build-arg`.

This would default to x86_64:

  ```Dockerfile
  ARG TARGET_CPU_ARCH="x86_64"
  RUN wget --no-hsts --quiet "https://github.com/conda-forge/miniforge/releases/download/${MINIFORGE_VERSION}/Miniforge3-${MINIFORGE_VERSION}-Linux-${TARGET_CPU_ARCH}.sh" -O /tmp/miniforge.sh && \ 
  ```

We could specify a different arch by specifying a different value for TARGET_CPU_ARCH:

  ```bash
  docker build --build-arg TARGET_CPU_ARG="aarch64" .
  ```

- https://docs.docker.com/engine/reference/commandline/build/#set-build-time-variables---build-arg
- https://vsupalov.com/docker-arg-env-variable-guide/

The new way to accomplish this (that only works with docker AFAIU?)  is with the `--platform` argument; which results in there being a number of env variablesset at build time:

  ```bash
  export DOCKER_BUILDKIT=1
  docker build --platform linux/amd64
  docker build --platform linux/arm64
  docker build --platform=linux/amd64,linux/arm64,linux/arm/v7 .
  docker build --platform=darwin .
  ```

  ```Dockerfile
  RUN wget --no-hsts --quiet "https://github.com/conda-forge/miniforge/releases/download/${MINIFORGE_VERSION}/Miniforge3-${MINIFORGE_VERSION}-Linux-${TARGETARCH}.sh" -O /tmp/miniforge.sh && \ 
  ```

- https://github.com/docker/buildx/#building-multi-platform-images
- https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope
- https://github.com/moby/moby/issues/32487
  - https://github.com/moby/buildkit/pull/499/files#diff-3e5af3861d799be5e17a086b5243b0fcR49
    - http://sujayopillai.com/2019/01/02/docker-platform-args.html has a Dockerfile that `echo`s each of the platform args but not their values

It looks like TARGETARCH is not quite what we're looking for in order to get `x86_64 | aarch64 | ppe64le` for the miniforge filename; so we may need a conditional on TARGETARCH or one of the other variables: https://github.com/moby/buildkit/pull/499#issuecomment-620183416

**Comment by isuruf** [comment]
`uname -m` works

**Comment by westurner** [comment]
I'm assuming that `uname -m` is something like `$BUILDARCH`; which isn't what we want when cross-compiling?

**Comment by martin-g** [comment]
Is there still an interest to merge `linux-anvil-cos7-x86_64`, `linux-anvil-aarch64` and `linux-anvil-ppc64le`  into a multiarch image ?
Currently this is a blocker for https://github.com/bioconda/bioconda-containers/issues/63 to provide multiarch image too.
I'd be happy to help here if there is an interest (and chance to be merged)!

**Comment by martin-g** [comment]
Here is PR in my fork that builds a multi-arch image for `linux-anvil-cos7`, i.e. CentOS 7 for AMD64, ARM64 and PPC64LE - https://github.com/martin-g/docker-images/pull/1/files
Here is the registry with the image: https://github.com/martin-g/docker-images/pkgs/container/condaforge%2Flinux-anvil-cos7

**Comment by jakirkham** [comment]
Thanks for your interest Martin! 🙏

Think what we would want to do is create a GHA step after the existing Docker builds that pulled the different architecture builds into multiarch images that also get pushed. IOW we could use [`docker manifest`]( https://docs.docker.com/engine/reference/commandline/manifest/#manifest-create ) to combine them

That way users can still retrieve images under the existing names, we still are able to parallelize/separate arch builds, and the end result maps nicely on to our existing images

Am curious to hear your thoughts on this approach 🙂

**Comment by martin-g** [comment]
Worked like a charm: 
PR: https://github.com/martin-g/docker-images/pull/3/files
Docker registry: https://hub.docker.com/repository/docker/mtgrigorov/linux-anvil-cos7-manifest/general

I did it in a separate workflow file because I didn't want to wait for the whole build of all images in ci.yaml, but the build job must be in the ci.yaml and should declare a dependency on the `build` job.

**Comment by jakirkham** [comment]
Yep Marcel think we are proposing the same thing 🤝, which it appears Martin has done 👍

Given your working experiment Martin, would you be interested in submitting a PR? 🙂

**Comment by martin-g** [comment]
> Given your working experiment Martin, would you be interested in submitting a PR?

Sure! I was waiting for confirmation that this is what is preferred!

**Comment by martin-g** [comment]
You can see an almost finished WIP at https://github.com/martin-g/docker-images/pull/4/files
At the moment it uses my accounts for DockerHub and Quay.io to fully test the jobs and deploys the images and manifests even for non-main branch, again only for testing!
Anyone is welcome to give feedback directly in the fork PR!
Once all is green I will create a PR here!
